File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

bool visit[205];
int cnt = 0, N;
vector<vector<int>> c;

void bfs(int i){
    cnt++;
    queue<int> q;
    q.push(i); visit[i] = 1;
    
    while(!q.empty()){
        int x = q.front(); q.pop();
        for(int y = 0; y < N; y++){
            if(x == y || visit[y]) continue;
            if(c[x][y] == 1) {
                visit[y] = 1;
                q.push(y);
            }
        }
    }
}


int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    c = computers;
    N = n;
    
    for(int i = 0; i < n; i++){
        if(!visit[i]) bfs(i); 
    }
    
    return cnt;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    int answer = 0, N = 0;
    boolean[] v;
    
    public int solution(int n, int[][] computers) {
        v = new boolean[n];
        N = n;
        
        for(int i = 0; i < n; i++) if(!v[i]) bfs(i, computers);
        
        return answer;
    }
    
    public void bfs(int x, int[][] computers){
        answer++;
        
        Queue<Integer> q = new LinkedList<>();
        q.add(x);
        v[x] = true;
        
        while(!q.isEmpty()){
            int cur = q.poll();
            for(int i = 0; i < N; i++) {
                if(cur == i || v[i] || computers[cur][i] != 1) continue;
                v[i] = true;
                q.add(i);
            } 
        }
    }
}